### DISCUSSION TITLE
Does Next.js use virtual DOM to make site interactable?

### DISCUSSION THREAD

**Discussion by mattp0123** [vercel/next.js#36990] [Help] [answered]
I really like Next.js. The dev experience for it is so good.
Today I was wondering, does Next.js compile UI code for manipulating the real DOM directly like Svelte or Solid.js?

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Next.js is a React framework, so it does what React does, which is collect updates from classes and function components, and compare whether or not those updates are correctly reflected in the DOM. If they are not, React updates the DOM accordingly.

One thing to differentiate Next.js from CRA for example, is that, Next.js uses React's dehydrated rendering on the Server, and now in React 18 it uses, a pipeable stream (or at least it will) which is given by React 18 in order to stream chunks of dehydrated components to the client. I am being a bit lazy here with the terms.

On the client, this dehydrated HTML string is hydrated and event listeners are added to the DOM, roughly speaking. When compared to Client Side Rendering frameworks, in Next.js the Time To Visible metric is very short, and the Time To Interactive is roughly equal.
